### PR TITLE
Basic custom llama smollm2

### PR DESCRIPTION
**Pull request by galopyz**
This PR adds a custom llama model for smollm2-135m model. The model loads weights from hugging face and trains with llm foundry. By using `simple_train_smollm2.py` it trains on c4 dataset using the `scripts/train/yamls/pretrain/custom_smollm2-135m.yaml`.  It also uses a callback to generate text sample before each eval.

I created the model mostly from Hugging Face llama code as it gave me more numerical stability than Rashcka's implementation. So, I added Hugging Face and EleutherAI's license in my llama code.

It currently uses sdpa, but I plan on switching to flash attention varlen when I add ModernBert unpadding. 
It still needs PEFT, KV cache, and modal support.

### CODE REVIEW

**Review by vishalbakshi — APPROVED** [review]
Reviewed and approved! Custom Llama model weights, per-layer and per-module outputs match the HF implementation as is shown [in this Colab](https://colab.research.google.com/drive/1sVcMLPj0D7JJCWoSi0UlX8nJO0NlTutm?usp=sharing). I also trained both the custom model and HF model (SmolLM2-135M) for 1000 batches of the avelinapythonedu 100k split and the loss curves are identical:

image
